photon-ml/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.function
+
+import breeze.linalg.Vector
+import org.apache.spark.broadcast.Broadcast
+
+import com.linkedin.photon.ml.normalization.NormalizationContext
+
+/**
+ * Trait for an objective function with L2 regularization.
+ */
+trait L2Regularization extends ObjectiveFunction {
+  protected var l2RegWeight: Double = 0D
+
+  /**
+   * Getter
+   *
+   * @return The L2 regularization weight
+   */
+  def l2RegularizationWeight: Double = l2RegWeight
+
+  /**
+   * Setter
+   *
+   * @param newRegWeight The new L2 regularization weight
+   */
+  protected[ml] def l2RegularizationWeight_=(newRegWeight: Double): Unit = l2RegWeight = newRegWeight
+
+  /**
+   * Compute the value of the function with L2 regularization over the given data for the given model coefficients.
+   *
+   * @param input The given data over which to compute the objective value
+   * @param coefficients The model coefficients used to compute the function's value
+   * @param normalizationContext The normalization context
+   * @return The computed value of the function
+   */
+  abstract override protected[ml] def value(
+    input: Data,
+    coefficients: Coefficients,
+    normalizationContext: Broadcast[NormalizationContext]): Double =
+    super.value(input, coefficients, normalizationContext) + l2RegValue(convertToVector(coefficients))
```

Comment:
Why this function has implementation but still being abstract?

Do we need to make it abstract with the implementations, or do we need an actual implementation for it?